partner_validate_vat/models/res_partner.py

```diff
@@ -0,0 +1,29 @@
+# coding: utf-8
+# Copyright 2017 Vauxoo (https://www.vauxoo.com) <info@vauxoo.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import models, api, fields, _
+from openerp.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    vat = fields.Char(copy=False)
+
+    @api.constrains('vat', 'country_id')
+    def _check_vat_uniqueness(self):
+        """ Check that the vat is unique in the level
+            where the partner in the tree
+        """
+
+        user_company = self.env.user.company_id
+
+        for partner in self:
```

Comment:
I didn't talk about add `commercial_partner_id` to `api.constrains()`
I did talk about add `self.mapped('commercial_partner_id').filtered('vat')` because the `commercial_partner_id` is a stored field.

To api.constrains you need use the field that trigger the commercial_partner_id